### PR TITLE
[Payment] Payment methods fee calculators

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | no
| License       | MIT
| Doc PR        | https://github.com/Sylius/Sylius-Docs/pull/301

Added fee calculators feature for payments, that gives possibility of charge additional fee for any payment method. Whole logic is based on shipment methods' calculators.

For now, I have implemented two basic calculators:
 - fixed fee calculator (i.e. €0.50)
 - percent fee calculator (i.e. 10% of all prize - including items, taxes and shipping)

### CODE REVIEW

**Comment by winzou** [comment]
Great feature. I just suggest you merge the 2 calculators in one with the 2 options fixed and percent. If a user have only a fixed fee, he will just put 0 in the percent field. Simpler, so better ;)

**Comment by Zales0123** [comment]
@winzou thank you for your appreciation!
I was wondering about your suggestion, but i think, that we should for now leave it as it is. Your idea can cause problems with calculators customization - now it can be done for each separately, which is good for potential customizer... And also nothing stops anybody from writing new calculator that coalesce this two fee calculator functionalities ;)
However I promise I will talk about this in my team and maybe we will find best solution.

File: features/frontend/checkout_payment.feature [hunk 1]
```diff
@@ -22,10 +22,10 @@ Feature: Checkout Payment
...
             | UK   | DHL Express |
           And the following payment methods exist:
-            | name        | gateway    | enabled |
-            | Credit Card | stripe     | yes     |
-            | PayPal      | paypal     | yes     |
-            | PayPal PRO  | paypal_pro | no      |
+            | name        | gateway    | enabled | calculator | calculator_configuration |
+            | Credit Card | stripe     | yes     | fixed      | amount: 10               |
+            | PayPal      | paypal     | yes     | fixed      | amount: 50               |
+            | PayPal PRO  | paypal_pro | no      | percent    | amount: 10               |
```
Comment (michalmarcinkowski) [inline features/frontend/checkout_payment.feature]: Always test correctness with ``` 0 ``` value. Please add these tests.

File: features/frontend/checkout_shipping.feature [hunk 2]
```diff
@@ -30,8 +30,8 @@ Feature: Checkout shipping
             | USA          | FedEx Premium | Flat rate  | Amount: 10000 | yes     |
             | UK + Germany | UPS Ground    | Flat rate  | Amount: 20000 | no      |
           And the following payment methods exist:
-            | name  | gateway | enabled |
-            | Dummy | dummy   | yes     |
+            | name  | gateway | enabled | calculator | calculator_configuration |
+            | Dummy | dummy   | yes     | fixed      | amount: 0.00             |
```
Comment (michalmarcinkowski) [inline features/frontend/checkout_shipping.feature]: Why float value?

File: features/frontend/checkout_payment.feature (hunk 1 above)
Comment (Zales0123) [inline features/frontend/checkout_payment.feature]: Unfortunately, ``getConfiguration`` method in ``DefaultContext`` cast 'amount' value to integer.
So for now i will leave this as it is and maybe change in another PR ;)

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! :+1:

File: src/Sylius/Component/Payment/Model/PaymentMethod.php [hunk 3]
```diff
@@ -61,6 +63,18 @@ class PaymentMethod implements PaymentMethodInterface
...
+     * FeeCalculator name
+     *
+     * @var string
+     */
+    protected $feeCalculator = DefaultFeeCalculators::FIXED;
+
+    /**
+     * @var array
+     */
+    protected $feeCalculatorConfiguration;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Payment/Model/PaymentMethod.php]:
```php
protected $feeCalculatorConfiguration = array();
```

File: src/Sylius/Component/Payment/Calculator/DefaultFeeCalculators.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class DefaultFeeCalculators
+{
+    /**
+     * Fixed fee calculator for payment
+     */
+    const FIXED = "fixed";
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Payment/Calculator/DefaultFeeCalculators.php]: Use single quotes:
```php
const FIXED = 'fixed';
```

Same below.

File: src/Sylius/Component/Payment/Model/PaymentMethod.php [hunk 5]
```diff
@@ -191,6 +205,42 @@ public function setEnvironment($environment)
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setFeeCalculator($feeCalculator)
+    {
+        $this->feeCalculator = $feeCalculator;
+
+        return $this;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Payment/Model/PaymentMethod.php]: We no longer do fluent interfaces.

Same below.

**Pull request merged**